libraries/AP_Motors/AP_MotorsMatrixTS.cpp

```diff
@@ -0,0 +1,104 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ *       AP_MotorsMatrixTS.cpp - tailsitters with multicopter motor configuration
+ */
+
+#include <AP_BattMonitor/AP_BattMonitor.h>
+#include <AP_HAL/AP_HAL.h>
+#include "AP_MotorsMatrixTS.h"
+
+extern const AP_HAL::HAL& hal;
+
+#define SERVO_OUTPUT_RANGE  4500
+
+// output a thrust to all motors that match a given motor mask. This
+// is used to control motors enabled for forward flight. Thrust is in
+// the range 0 to 1
+void AP_MotorsMatrixTS::output_motor_mask(float thrust, uint8_t mask, float rudder_dt)
+{
+    for (uint8_t i=0; i<AP_MOTORS_MAX_NUM_MOTORS; i++) {
+        if (motor_enabled[i]) {
+            int16_t motor_out;
+            if (mask & (1U<<i)) {
+                /*
+                    apply rudder mixing differential thrust
+                    copter frame roll is plane frame yaw as this only
+                    apples to either tilted motors or tailsitters
+                */
+                float diff_thrust = get_roll_factor(i) * rudder_dt * 0.5f;
+                thrust = constrain_float(thrust + diff_thrust, 0.0f, 1.0f);
+                motor_out = get_pwm_output_min() + (get_pwm_output_max()-get_pwm_output_min()) * thrust;
+            } else {
+                motor_out = get_pwm_output_min();
+            }
+            rc_write(i, motor_out);
+        }
+    }
+}
+
+void AP_MotorsMatrixTS::output_to_motors()
+{
+    // calls calc_thrust_to_pwm(_thrust_rpyt_out[i]) for each enabled motor
+    AP_MotorsMatrix::output_to_motors();
+
+    // also actuate control surfaces
+    SRV_Channels::set_output_scaled(SRV_Channel::k_aileron,  -_yaw_in * SERVO_OUTPUT_RANGE);
+    SRV_Channels::set_output_scaled(SRV_Channel::k_elevator, _pitch_in * SERVO_OUTPUT_RANGE);
+    SRV_Channels::set_output_scaled(SRV_Channel::k_rudder, _roll_in * SERVO_OUTPUT_RANGE);
+}
+
+void AP_MotorsMatrixTS::output_armed_stabilizing()
+{
+    // calculates thrust values _thrust_rpyt_out
+    AP_MotorsMatrix::output_armed_stabilizing();
+}
+
+void AP_MotorsMatrixTS::setup_motors(motor_frame_class frame_class, motor_frame_type frame_type)
+{
+    // remove existing motors
+    for (int8_t i=0; i<AP_MOTORS_MAX_NUM_MOTORS; i++) {
+        remove_motor(i);
+    }
+
+    bool success = false;
+
+    switch (frame_class) {
+
+        case MOTOR_FRAME_QUAD:
```

Comment:
why not just add this bit to the existing motor matrix? You could add a Quad tail-sitter frame class, I think the only other change is to push the outputs to plane, I think pulling them in as you had previously is a better way to do it. I think they are the only differences between this and the standard motor matrix?

Reply:
I like this because it separates quadplane support from pure multicopter stuff, and I think it will be easier to get Randy to approve it.  Having the new class also allows me to override output_motor_mask to change throttle behavior only for the new frame type.
I also plan to override output_armed_stabilizing to implement body-frame yaw...

Reply:
maybe we could work out a way to allow the old matrix to be used as well, a new frame class or parameter or something.  In for example a x frame quad the props would not be mounted on the wings so ailerons only for yawing would not work. Or if you just wanted normal size ones rather than making them extra big. And we get all the existing frame class of course.

Reply:
I think X and H types could easily be added to MatrixTS just by copy/paste in setup_motors. 
Not sure that's a good design approach though; adds a lot of structure for no good reason I can think of and reduces control authority too.  I'll add a commit which does that, but don't plan to build a real or RF8 model to test it.  
Also, I see that doing body-frame yaw will involve changing quadplane.cpp, since that's where the copter attitude controller is invoked. Not sure It's a good idea to make that any more complicated.  output_armed_stabilizing could be simplified if torque-based yaw is not in use, but there's probably not much to gain from that.
Have you seen the designs by Xcraft at xcraft.io? Looks like they started with a quad-plus frame, no elevons, then designed the X2, a tri config with elevons.  I wonder why they put motors on the wingtips instead of midspan.
They are also looking for an engineer with Ardupilot experience...

Reply:
If you add your new frames as a new frame types rather than switching based on the tailsit mask we can have both versions of each frame, and just be tail sitter or not based on the mask. With the new pulling method of controls we can use the copter matrix with no modifications.  
I looked in to body frame yawing very briefly a few months ago and came to the conclusion that at angles where there would be a noticeable difference you would quickly come up against Euler angle issues. The solution is to go to quaternions, I did try and do some reading up bit it gets very maths heavy. The attitude controller is set up for quaternions we would 'just' need to convert pilots inputs in to desired.  
Its a interesting design, cant really see what they get from the 3rd motor. Cant be used for forward flight and only helps pitching one way. Unless the CG is a 3rd of the way up towards it but i don't think that would work for forward flight, so in the hover the motor must be basically idling. They still need the big flaps to pitch up why not just use them to pitch down too.

Reply:
I agree that everything except changing the throttle behavior of output_motor_mask can be done without a new subclass. Let's see what advice @tridge has on which way to go.
Notice that the X2 top motor is less than half the power of each of the other two motors, so that probably gives us a hint of where the CG lies on the Z axis. I expect that a tri-motor arrangement will perform well in hover and also can be more efficient in FW flight too.  And I don't think a z-offset on the CG should be a problem in FW mode; will have to do an RF8 model to play with.